src/chrome/internal/sources/unresolvedSource.ts

```diff
@@ -0,0 +1,56 @@
+import { IResourceIdentifier } from './resourceIdentifier';
+import { ILoadedSource } from './loadedSource';
+import { SourceResolver } from './sourceResolver';
+
+/**
+ * VS Code debug protocol sends breakpoint requests with a path?: string; or sourceReference?: number; Before we can use the path, we need to wait for the related script to be loaded so we can match it with a script id.
+ * This set of classes will let us represent the information we get from either a path or a sourceReference, and then let us try to resolve it to a script id when possible.
+ */
+export interface IUnresolvedSource {
+    readonly sourceIdentifier: IResourceIdentifier;
+    isEquivalent(right: IUnresolvedSource): boolean;
+    tryResolving<R>(succesfulAction: (resolvedSource: ILoadedSource) => R, failedAction: (sourceIdentifier: IResourceIdentifier) => R): R;
+}
+
+abstract class UnresolvedSourceCommonLogic implements IUnresolvedSource {
+    public abstract tryResolving<R>(succesfulAction: (loadedSource: ILoadedSource) => R, failedAction: (identifier: IResourceIdentifier) => R): R;
+    public abstract get sourceIdentifier(): IResourceIdentifier;
+
+    public isEquivalent(right: IUnresolvedSource): boolean {
+        return this.sourceIdentifier.isEquivalent(right.sourceIdentifier);
+    }
+}
+
+// Find the related source by using the source's path
+export class SourceToBeResolvedViaPath extends UnresolvedSourceCommonLogic implements IUnresolvedSource {
+    public tryResolving<R>(succesfulAction: (resolvedSource: ILoadedSource) => R, failedAction: (sourceIdentifier: IResourceIdentifier) => R) {
+        return this._sourceResolver.tryResolving(this.sourceIdentifier, succesfulAction, failedAction);
+    }
+
+    public toString(): string {
+        return `Resolve source via #${this.sourceIdentifier}`;
+    }
+
+    constructor(public readonly sourceIdentifier: IResourceIdentifier, private readonly _sourceResolver: SourceResolver) {
+        super();
+    }
+}
+
+// This source was already loaded, so we store it in this class
+export class SourceAlreadyResolvedToLoadedSource extends UnresolvedSourceCommonLogic implements IUnresolvedSource {
```

Comment:
I don't understand why there is a "SourceAlreadyResolved" which is an "IUnresolvedSource", where would this be used?

Reply:
You are right, it doesn't make sense.

I'd love your help coming up with some better names.

We use SourceAlreadyResolvedToLoadedSource when the client sends a sourceReference parameter. We use SourceToBeResolvedViaPath when the client sends a path parameter.

Maybe I should rename IUnresolvedSource to ISourceThatMayBeUnresolved?

Suggestions are welcome....

Reply:
If it's sometimes unresolved and sometimes resolved then it seems like it's just a "Source".

Reply:
Renamed to ISource...